File: Python/bsseq_analysis.py
"""
Bisulfite Sequencing Analysis in Python
Author: Dr. Divya Mishra, Ph.D.
Purpose: Load Bismark methylation calls, visualize CpG methylation and DMRs
"""

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# --------------------------
# 1. Load methylation data
# --------------------------
sample1 = pd.read_csv('results/methylation/sample1_CpG.txt', sep='\t')
sample2 = pd.read_csv('results/methylation/sample2_CpG.txt', sep='\t')

# Merge by genomic position
merged = pd.merge(sample1, sample2, on=['chr','start','end'], suffixes=('_s1','_s2'))

# --------------------------
# 2. Visualize methylation distribution
# --------------------------
plt.figure(figsize=(8,6))
sns.histplot(merged['meth_percent_s1'], color='blue', label='Sample1', kde=True, stat="density")
sns.histplot(merged['meth_percent_s2'], color='red', label='Sample2', kde=True, stat="density")
plt.xlabel('Methylation Percentage')
plt.ylabel('Density')
plt.title('CpG Methylation Distribution')
plt.legend()
plt.savefig('results/figures/methylation_distribution_python.png', dpi=300)
plt.close()

# --------------------------
# 3. Optional: DMR Scatter Plot
# --------------------------
merged['diff'] = merged['meth_percent_s2'] - merged['meth_percent_s1']

plt.figure(figsize=(8,6))
sns.scatterplot(x='meth_percent_s1', y='meth_percent_s2', data=merged, alpha=0.3)
plt.xlabel('Sample1 Methylation (%)')
plt.ylabel('Sample2 Methylation (%)')
plt.title('CpG Methylation Comparison')
plt.savefig('results/figures/DMR_scatter.png', dpi=300)
plt.close()

File: R/bsseq_analysis.R
# =================================================================
# Bisulfite Sequencing Analysis in R
# Author: Dr. Divya Mishra, Ph.D.
# Purpose: Load Bismark output, perform DMR detection, and visualize
# =================================================================

# Load libraries
library(methylKit)
library(GenomicRanges)
library(ggplot2)

# --------------------------
# 1. Load methylation data
# --------------------------
file.list <- list(
  "results/methylation/sample1_CpG.txt",
  "results/methylation/sample2_CpG.txt"
)

sample.id <- c("sample1","sample2")
treatment <- c(0,1)  # 0=control, 1=treated

myobj <- methRead(file.list,
                  sample.id=sample.id,
                  assembly="hg38",
                  treatment=treatment,
                  context="CpG",
                  pipeline="bismarkCoverage")

# --------------------------
# 2. Filtering and normalization
# --------------------------
filtered <- filterByCoverage(myobj, lo.count=10, hi.perc=99.9)
normalized <- normalizeCoverage(filtered)

# --------------------------
# 3. Unite samples (common CpG sites)
# --------------------------
meth <- unite(normalized, destrand=TRUE)

# --------------------------
# 4. Differential methylation analysis
# --------------------------
myDiff <- calculateDiffMeth(meth, covariates=NULL, overdispersion="MN")

# Extract DMRs
dmr <- getMethylDiff(myDiff, difference=25, qvalue=0.01)
write.csv(dmr, "results/DMRs/DMR_results.csv")

# --------------------------
# 5. Visualization
# --------------------------
# Methylation distribution
pdf("results/figures/methylation_distribution.pdf")
getMethylationStats(meth[[1]], plot=TRUE, both.strands=FALSE)
dev.off()

# DMR heatmap
pdf("results/figures/DMR_heatmap.pdf")
clusterSamples(myDiff, dist="correlation", method="ward", plot=TRUE)
dev.off()
